modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultData.java

```diff
@@ -0,0 +1,500 @@
+
+            return null;
+        }
+
+        // 2. Read body.
+        // Preallocate double row size to minimize possible allocations for other rows.
+        if (readBuff == null || readBuff.capacity() < size)
+            readBuff = ByteBuffer.allocate(size * 2);
+
+        readBuff.clear();
+        readBuff.limit(size);
+
+        readFromFile(readBuff);
+
+        readBuff.flip();
+
+        Data buff = Data.create(null, readBuff.array(), true);
+
+        IgniteBiTuple<ValueRow, Value[]> row = new IgniteBiTuple<>();
+
+        Value rowKey = buff.readValue();
+
+        if (rowKey == ValueNull.INSTANCE)
+            row.set1(null);
+        else
+            row.set1((ValueRow)rowKey);
+
+        Value[] rowVal = new Value[colCnt - 1];
+
+        for (int i = 0; i < colCnt - 1; i++)
+            rowVal[i] = buff.readValue();
+
+        row.set2(rowVal);
+
+        return row;
+
+    }
+
+    /**
+     * Reads data from file into the buffer.
+     *
+     * @param buff Buffer.
+     */
+    private void readFromFile(ByteBuffer buff) {
+        try {
+            fileIo.readFully(buff);
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Writes buffer to file.
+     *
+     * @param buff Buffer.
+     * @return Bytes written.
+     */
+    private int writeToFile(Data buff) {
+        try {
+            ByteBuffer byteBuff = ByteBuffer.wrap(buff.getBytes());
+
+            byteBuff.limit(buff.length());
+
+            fileIo.writeFully(byteBuff);
+
+            return byteBuff.limit();
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets arbitrary file position.
+     *
+     * @param pos Position to set.
+     */
+    private void setFilePosition(long pos) {
+        try {
+            fileIo.position(pos);
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to reset the spill file.", e);
+        }
+    }
+
+    /**
+     * @return Current absolute position in the file.
+     */
+    private long currentFilePosition() {
+        try {
+            return fileIo.position();
+        }
+        catch (IOException e) {
+            close();
+
+            throw new IgniteException("Failed to access the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets position in file on the beginning.
+     */
+    void rewindFile() {
+        setFilePosition(0);
+    }
+
+    /**
+     * @return File chunks.
+     */
+    Collection<Chunk> chunks() {
+        return chunks;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() {
+        U.closeQuiet(fileIo);
+        file.delete();
+
+        U.closeQuiet(hashIdx);
```

Comment:
Maybe better to surround this code by try-finally because first lines of code can throw IoExeption.
```
try {
        U.closeQuiet(fileIo);
        file.delete();
} finally {
      U.closeQuiet(hashIdx);
}
```

Reply:
Agree. I moved 'U.closeQuiet(hashIdx)' to the second line right after 'U.closeQuiet(fileIo)' which can not throw 'IOException".